src/CacheControl/CacheControlServiceProvider.php

```diff
@@ -0,0 +1,34 @@
+<?php
+
+namespace Nuwave\Lighthouse\CacheControl;
+
+use Illuminate\Contracts\Events\Dispatcher;
+use Illuminate\Support\ServiceProvider;
+use Nuwave\Lighthouse\Events\EndRequest;
+use Nuwave\Lighthouse\Events\RegisterDirectiveNamespaces;
+
+class CacheControlServiceProvider extends ServiceProvider
+{
+    public function register(): void
+    {
+        $this->app->singleton(CacheControl::class);
+    }
+
+    public function boot(Dispatcher $dispatcher, CacheControl $cacheControl): void
+    {
+        $dispatcher->listen(
+            RegisterDirectiveNamespaces::class,
+            static function (): string {
+                return __NAMESPACE__;
+            }
+        );
+
+        $dispatcher->listen(
+            EndRequest::class,
+            function (EndRequest $request) use ($cacheControl): void {
+                $request->response->setCache($cacheControl->makeHeaderOptions());
+                app()->forgetInstance(CacheControl::class);
```

Comment:
Will a new singleton be instantiated for new requests? This might happen when using `Octane`.

Reply:
With data providers in tests, I needed to remove the instance otherwise tests failed. After the previous review changes, tests passed even without forgetInstance. I don't know what is happened, so pls decide to do forgetInstance or not.

Reply:
My concern was that after this call, the class may no longer be instantiated as a singleton. I tested the behaviour of the container though, and it does behave the way we need it to:

```
>>> class Foo {}
>>> app()->singleton(Foo::class)
=> null
>>> app(Foo::class)
=> Foo {#4859}
>>> app(Foo::class)
=> Foo {#4859}
>>> app()->forgetInstance(Foo::class)
=> null
>>> app(Foo::class)
=> Foo {#4857}
>>> app(Foo::class)
=> Foo {#4857}
```

We should actually keep this, it is crucial for deployments that have long running server processes.